### ISSUE TITLE
Hooking instruction has missing parameter

### ISSUE THREAD

**Issue by smeng9** [AlexAltea/unicorn.js#20]
It seems when I try to perform the hooking of a single instruction, I have no where to pass which instruction (for example syscall) that is being hooked in the hook_add function.

https://github.com/AlexAltea/unicorn.js/blob/dedbfcac3b402cdc363a3c03143629f6534a10c3/src/unicorn-wrapper.js#L185

Thanks

**Comment by AlexAltea** [comment]
It seems Unicorn's `uc_hook_add` function uses variadic arguments, based on the hook type.
Not sure whether Emscripten supports it, but we can try it:

I'd propose these changes:
```diff
-         this.hook_add = function (type, user_callback, user_data, begin, end) {
+         this.hook_add = function (type, user_callback, user_data, begin, end, extra) {
```

```diff
              // Wrap callback
+             var extra_types = [];
+             var extra_values = [];
              switch (type) {
                  case uc.HOOK_INSN:
+                     extra_types = ['number']
+                     extra_values = [extra]
```

```diff
              var ret = MUnicorn.ccall('uc_hook_add', 'number',
                  ['pointer', 'pointer', 'number', 'pointer', 'pointer',
-                     'number', 'number', 'number', 'number'],
+                     'number', 'number', 'number', 'number'] + extra_types,
                  [handle, hook_ptr, type, callback_ptr, 0,
-                     begin, 0, end, 0]
+                     begin, 0, end, 0] + extra_values
              );
```

Can you quickly test it and see if it works?

**Comment by smeng9** [comment]
Hi, I believe your method should work, because a lot of other bindings using this type of variable argument method.

However, there are still some issue with my Emscripten setup. When I recompiled the code no hooks (even of other types) can be catched, and this breaks my previous working code. I have checked the return handle of uc_hook_add and the value is 0. ~~I think there are some regression on Emscripten version. I will try to install an even older version of Emscripten than 1.37.23.~~

**Comment by smeng9** [comment]
EDIT: we should not use + for arrays, we should use the concat method

**Comment by smeng9** [comment]
The variadic argument indeed created some issue for me. I still did not get the instruction hook to work.

I searched the issue list of Emscripten and I believe they provided some patches in 1.38+ which may break our build script.

Currently I'm not sure which route may be easier:
1. Patch in build script for higher versions of Emsdk
2. Patch the c code that replaces varargs function with a normal function
I would like to listen to your opinion before picking on either ones since you have more experience working on this library

**Comment by AlexAltea** [comment]
> we should not use + for arrays, we should use the concat method

Sure! I don't regularly write JavaScript, so I forget the syntax quirks. 😄 

> Currently I'm not sure which route may be easier: [...]  I would like to listen to your opinion before picking on either ones.

I believe the 1st option is better:
- If something is broken, we should address the root cause and update EMSDK (and hence the build scripts).
- Patching C code is incredibly difficult and hard to read/debug (due to regex and macros). It feels also a last-resort measure.
- Changes needed to adapt to newer EMSDK versions have been (so far) quite straightforward in my opinion.

**Comment by smeng9** [comment]
After testing Emscripten ccall with simple varargs function (e.g. adding a bunch of numbers), even the latest version still did not produce correct result. Seems the patch to Emscripten https://github.com/emscripten-core/emscripten/pull/5684 is tagged won't fix and never gets merged. I will try to bring the issue to Emscripten again.

Now I have an awkward fix that uses capstone engine to read and decode each memory to see which instruction is which and stop and resume on that address. That's too much wrapping and slows down the code significantly.

I feel I have to resort to patching the C code. The hood_add function seems is an exported function of unicorn library. There are not too many code depend on that particular function signature besides some test cases. Or we can just add a new function that wraps the varargs function in C.

**Comment by AlexAltea** [comment]
This should have been solved by @smeng9 in #24.
